Add support for Vobsub subtitles

wischnow opened the pull request:
Hello,

this add support for Vobsub subtitles.

An example file can be found here: https://www.w9y.de/sample.mkv
This is based on this file: https://samples.ffmpeg.org/sub/largeres_vobsub.mkv

The only difference is that I used "mkvmerge -o sample.mkv largeres_vobsub.mkv" to copy the tracks into a new MKV file because the original uses ContentCompAlgo == 0, which is not supported by Exoplayer.

icbaker commented:
Thanks for the contribution! In order to merge this we need a (very) small test video containing at least one vobsub subtitle that we can use to write a playback test to verify everything works (and continues to work as we change things in future). 

This asset needs to be suitably licensed for submission to this repo. The `sample.mkv` file you've linked to is too large (40MB), and doesn't look suitably licensed. The easiest way to generate these test assets is usually to craft them yourself, maybe by modifying one of our [existing MKV assets](https://github.com/androidx/media/tree/main/libraries/test_data/src/test/assets/media/mkv).

I had a quick look around and didn't find an easy way to synthesise vobsub subtitles. Please can you take a look into this. Once you have the small asset with a vobsub track added to this PR, you can probably just add the file to the list of assets in [`MkvPlaybackTest`](https://github.com/androidx/media/blob/main/libraries/exoplayer/src/test/java/androidx/media3/exoplayer/e2etest/MkvPlaybackTest.java) and run the test with [`DumpFIleAsserts.DUMP_FILE_ACTION`](https://github.com/androidx/media/blob/main/libraries/test_utils/src/main/java/androidx/media3/test/utils/DumpFileAsserts.java) hard-coded locally to `WRITE_TO_LOCAL` (this should produce the golden file dump output).

wischnow commented:
Hello and thanks for your reply!

While searching for a better sample file or how to build one I just saw that ffmpeg can convert other image-based subtitle formats to Vobsub/DVDsub. And since Exoplayer does support PGS and DVBsub subtitles, which are image-based, I thought I'd just use one of your sample files for those and convert them to Vobsub.

But I can't find a sample file with PGS or DVBsub subtitles in the repository or in the old ExoPlayer repository.

Is that correct, or am I missing something? If I could get one of the files you used for those subtitle formats, I can easily create one with Vobsub subtitles.



P.S.: It appears that the Vobsub format ffmpeg outputs is a bit weird in that it doesn't contain the "size:..." line in the embedded IDX file. Never seen that before. My code needs that to be able to calculate the relative positions required by Cue.Builder(). At least I didn't see a way how a subtitle parser could get access to the viewport size. But the output of ffmpeg can easily be fixed by adding the "size:..." line by hand (mkvextract, an editor, and mkvmerge).

icbaker commented:
> Is that correct, or am I missing something?

That's correct, we don't have test assets for these subtitle formats either, and this introduces a maintenance burden when we want to make changes (because we have less confidence that this logic will still work) - which is why I'm afraid we now require test assets for new subtitle formats.

----

Some quick searching [suggests that tsMuxer might be able to convert from SRT to an image-based format](https://www.reddit.com/r/ffmpeg/comments/da0hxp/can_ffmpeg_create_dvb_subtitles_from_text/), which maybe ffmpeg could then convert to vobsub? I haven't tried this (or ever used this tool).

Or if there's a way to just mux arbitrary images in as subtitles, you could use some of our existing [test images](https://github.com/androidx/media/tree/main/libraries/test_data/src/test/assets/media/jpeg) (there are other formats too).

wischnow commented:
Hello again (and a happy new year),

thanks for pointing me at tsmuxer. This worked. I took the existing `sample_with_srt.mkv` and converted the subtitle to Vobsub format (via tsmuxer and ffmpeg). This I then added to my fork and also added the file to `MkvPlaybackTest.java`, as suggested.

icbaker commented on libraries/extractor/src/main/java/androidx/media3/extractor/text/vobsub/VobsubParser.java:
```diff
@@ -0,0 +1,351 @@
+        parseRun(bitBuffer, run);
+
+        l = run.length;
+        if (l > width - x) l = width - x;
```
nit: simplify this to `length = min(run.length, width - x)`

wischnow replied:
Fixed

icbaker commented on libraries/extractor/src/main/java/androidx/media3/extractor/text/vobsub/VobsubParser.java:
```diff
@@ -0,0 +1,351 @@
+          bitmapData[i++] = run.color;
+          x++;
+          l--;
+        }
```
I think you can use `Arrays.fill` here

wischnow replied:
Fixed

icbaker commented on libraries/extractor/src/main/java/androidx/media3/extractor/text/vobsub/VobsubParser.java:
```diff
@@ -0,0 +1,351 @@
+    private void parseRleData(int[] bitmapData, ParsableBitArray bitBuffer, int y) {
+      int x = 0;
+      int l = 0;
+      int i = y * width;
```
what does `i` represent here? a longer name might make this more clear

wischnow replied:
Fixed

icbaker commented on libraries/extractor/src/main/java/androidx/media3/extractor/text/vobsub/VobsubParser.java:
```diff
@@ -0,0 +1,351 @@
+
+    private void parseRun(ParsableBitArray bitBuffer, Run run) {
+      int v = 0;
+      int t = 1;
```
same here, please give these longer, clearer names

wischnow replied:
Fixed

icbaker commented on libraries/extractor/src/main/java/androidx/media3/extractor/text/vobsub/VobsubParser.java:
```diff
@@ -0,0 +1,351 @@
+        t <<= 2;
+      }
+      run.color = colors[v & 3];
+      if (v < 4) run.length = width;
```
nit: consider using a ternary operator for this:

```
run.length = v < 4 ? width : v >> 2;
```

wischnow replied:
Fixed

wischnow commented on libraries/extractor/src/main/java/androidx/media3/extractor/text/vobsub/VobsubParser.java:
```diff
@@ -0,0 +1,351 @@
+        buffer.reset(inflatedBuffer.getData(), inflatedBuffer.limit());
+      } // else assume data is not compressed.
+    }
+  }
```
Yes, I had thought about factoring this out, but didn't want to mess with other people's code if not needed. I have now added patches for this: one adding it to Util, one changing PgsParser and one with this and various other changes for VobsubParser.

wischnow commented on libraries/extractor/src/main/java/androidx/media3/extractor/text/vobsub/VobsubParser.java:
```diff
@@ -0,0 +1,351 @@
+    public void parseIdx(String idx) {
+      for (String line : idx.trim().split("\\r?\\n")) {
+        if (line.startsWith("palette: ")) {
+          String[] values = line.substring(9).split(",");
```
Fixed

wischnow commented on libraries/extractor/src/main/java/androidx/media3/extractor/text/vobsub/VobsubParser.java:
```diff
@@ -0,0 +1,351 @@
+      for (String line : idx.trim().split("\\r?\\n")) {
+        if (line.startsWith("palette: ")) {
+          String[] values = line.substring(9).split(",");
+          int l = values.length;
```
Fixed

wischnow commented on libraries/extractor/src/main/java/androidx/media3/extractor/text/vobsub/VobsubParser.java:
```diff
@@ -0,0 +1,351 @@
+    public void parseIdx(String idx) {
+      for (String line : idx.trim().split("\\r?\\n")) {
+        if (line.startsWith("palette: ")) {
+          String[] values = line.substring(9).split(",");
```
Fixed

wischnow commented on libraries/extractor/src/main/java/androidx/media3/extractor/text/vobsub/VobsubParser.java:
```diff
@@ -0,0 +1,351 @@
+      }
+    }
+
+    private int parseColor(String value) {
```
Fixed

wischnow commented on libraries/extractor/src/main/java/androidx/media3/extractor/text/vobsub/VobsubParser.java:
```diff
@@ -0,0 +1,351 @@
+      }
+    }
+    if (cue != null) {
+      cues.add(cue);
```
Fixed

wischnow commented on libraries/extractor/src/main/java/androidx/media3/extractor/text/vobsub/VobsubParser.java:
```diff
@@ -0,0 +1,351 @@
+  public static final @CueReplacementBehavior int CUE_REPLACEMENT_BEHAVIOR =
+      Format.CUE_REPLACEMENT_BEHAVIOR_REPLACE;
+
+  private static final int DEFAULT_DURATION = 5000000;
```
Not changed because I've seen this very often, that Vobsub subtitles stay on screen if no duration is given. In fact, it led me to find a bug in the Exoplayer code which I reported beginning of December and which has since been fixed (see https://github.com/androidx/media/issues/1939#issue-2712258130).

wischnow commented on libraries/extractor/src/main/java/androidx/media3/extractor/text/vobsub/VobsubParser.java:
```diff
@@ -0,0 +1,351 @@
+      hasColors = false;
+      hasPosition = false;
+      hasDataOffsets = false;
+      palette = null;
```
Fixed

wischnow commented on libraries/extractor/src/main/java/androidx/media3/extractor/text/vobsub/VobsubParser.java:
```diff
@@ -0,0 +1,351 @@
+        } else if (line.startsWith("size: ")) {
+
+          // NOTE: we need this line to calculate the relative positions
+          //       and size required by Cue.Builder() below.
```
No, I really mean the place where Cue.Builder() is created (with the latest changes, this is line 288).

wischnow commented on libraries/extractor/src/main/java/androidx/media3/extractor/text/vobsub/VobsubParser.java:
```diff
@@ -0,0 +1,351 @@
+  private static final int CMD_ALPHA   = 4;
+  private static final int CMD_AREA    = 5;
+  private static final int CMD_OFFSETS = 6;
+  private static final int CMD_END     = 255;
```
Fixed

wischnow commented on libraries/extractor/src/main/java/androidx/media3/extractor/text/vobsub/VobsubParser.java:
```diff
@@ -0,0 +1,351 @@
+    }
+
+    private void parseControl(ParsableByteArray buffer, int end) {
+      int t, d0, d1, d2;
```
Fixed

wischnow commented on libraries/extractor/src/main/java/androidx/media3/extractor/text/vobsub/VobsubParser.java:
```diff
@@ -0,0 +1,351 @@
+      buffer.setPosition(pos);
+
+      int end = buffer.readUnsignedShort();
+      parseControl(buffer, end);
```
That's already set and used for the whole buffer in parse().

wischnow commented on libraries/extractor/src/main/java/androidx/media3/extractor/text/vobsub/VobsubParser.java:
```diff
@@ -0,0 +1,351 @@
+            colors[3] = getColor(d0 >> 4);
+            colors[2] = getColor(d0 & 0xf);
+            colors[1] = getColor(d1 >> 4);
+            colors[0] = getColor(d1 & 0xf);
```
Wouldn't that be overkill? We're just getting two bytes and separating the nibbles here.

wischnow commented on libraries/extractor/src/main/java/androidx/media3/extractor/text/vobsub/VobsubParser.java:
```diff
@@ -0,0 +1,351 @@
+    private int planeWidth;
+    private int planeHeight;
+    private int[] colors;
+    private int x0, y0, x1, y1, width, height;
```
Fixed (using Rect).

wischnow commented on libraries/extractor/src/main/java/androidx/media3/extractor/text/vobsub/VobsubParser.java:
```diff
@@ -0,0 +1,351 @@
+            y0 = (d0 << 4) | (d1 >> 4);
+            y1 = ((d1 & 0xf) << 8) | d2;
+            width = x1 - x0 + 1;
+            height = y1 - y0 + 1;
```
Fixed

wischnow commented on libraries/extractor/src/main/java/androidx/media3/extractor/text/vobsub/VobsubParser.java:
```diff
@@ -0,0 +1,351 @@
+          .build();
+    }
+
+    private void parseRleData(int[] bitmapData, ParsableBitArray bitBuffer, int y) {
```
Fixed

wischnow commented on libraries/extractor/src/main/java/androidx/media3/extractor/text/vobsub/VobsubParser.java:
```diff
@@ -0,0 +1,351 @@
+
+    private void parseRleData(int[] bitmapData, ParsableBitArray bitBuffer, int y) {
+      int x = 0;
+      int l = 0;
```
Fixed

icbaker commented on libraries/extractor/src/main/java/androidx/media3/extractor/text/vobsub/VobsubParser.java:
```diff
@@ -0,0 +1,363 @@
+    }
+
+    private boolean parseControlAlpha(ParsableByteArray buffer) {
+      if (buffer.bytesLeft() < 2) return false;
```
should we also exit early if `hasColors == false`? I assume (?) that the colors must be defined before the alpha. If that's not the case, then the `parseControlColors()` code above looks problematic, because it will overwrite any alpha data stored in the `colors` array.

wischnow replied:
That makes sense. I guess I didn't catch that because I've only ever seen alpha being set after the colors have been set. Of course it would be more robust to collect the alphas in an array and then call the setAlpha()s after parseControl()

icbaker commented on libraries/extractor/src/main/java/androidx/media3/extractor/text/vobsub/VobsubParser.java:
```diff
@@ -0,0 +1,363 @@
+    /**
+     * Parse run-length encoded data into the {@code bitmapData} array. The
+     * subtitle bitmap is encoded in two blocks of interlaced lines, {@code y}
+     * gives the index of the starting line (0 or 1).
```
if `y` is always zero or one, what about making it a boolean?

wischnow replied:
Maybe I should have written the comment better: the function is called with y being 0 or 1, basically telling it whether it will decode the even or odd lines. Inside the function it will then increase y in the loop.

icbaker replied:
gotcha - I've made the parameter a boolean and made the `y` arithmetic an implementation detail inside the method.

icbaker commented on libraries/extractor/src/main/java/androidx/media3/extractor/text/vobsub/VobsubParser.java:
```diff
@@ -0,0 +1,363 @@
+    buffer = new ParsableByteArray();
+    inflatedBuffer = new ParsableByteArray();
+    cueBuilder = new CueBuilder();
+    cueBuilder.parseIdx(new String(initializationData.get(0), StandardCharsets.UTF_8));
```
I'm considering moving the `parseIdx` functionality into the `CueBuilder` constructor so that the `palette` field can be final and non-null.

wischnow replied:
That makes sense, too.

icbaker reviewed (commented):
I've made a few changes as part of preparing this for internal review, but also found a couple of questions I want your input on (see individual comments). I'm happy to make any resulting code changes, so feel free to just reply in the comments.

icbaker commented:
> thanks for pointing me at tsmuxer. This worked. I took the existing `sample_with_srt.mkv` and converted the subtitle to Vobsub format (via tsmuxer and ffmpeg). This I then added to my fork and also added the file to `MkvPlaybackTest.java`, as suggested.

Thanks for adding this test file - that's great you got `tsmuxer` working. Unfortunately when I ran the `MkvPlaybackTest` to try and generate an output dump file to add to this PR, it timed out:

```
java.util.concurrent.TimeoutException
	at androidx.media3.test.utils.robolectric.RobolectricUtil.runLooperUntil(RobolectricUtil.java:149)
	at androidx.media3.test.utils.robolectric.RobolectricUtil.runMainLooperUntil(RobolectricUtil.java:98)
	at androidx.media3.test.utils.robolectric.RobolectricUtil.runMainLooperUntil(RobolectricUtil.java:76)
	at androidx.media3.test.utils.robolectric.TestPlayerRunHelper$PlayerRunResult.runUntil(TestPlayerRunHelper.java:307)
	at androidx.media3.test.utils.robolectric.TestPlayerRunHelper$PlayerRunResult.untilState(TestPlayerRunHelper.java:122)
	at androidx.media3.test.utils.robolectric.TestPlayerRunHelper.runUntilPlaybackState(TestPlayerRunHelper.java:608)
	at androidx.media3.exoplayer.e2etest.MkvPlaybackTest.test(MkvPlaybackTest.java:82)
```

Did you get the test running successfully with this file?

wischnow commented:
I had uploaded the sample file to [https://www.w9y.de/sample.mkv](url), added that to the `media.exolist.json` file of the demo app and tested it with that. I just tried it again and it still works, no error messages in the log, including no timeout (debug compile, btw.).

Hm, weird. How does one run those tests?

wischnow commented:
Correction: If I pull the latest version with your changes, the demo player doesn't display the subtitle...

wischnow commented:
And before I forget this again: from that conversion with tsmuxer I still have to sample file with PGS subtitles. Since this PR is about Vobsub/DVDsub I'm reluctant to add it here. You can get it here, if you're interested to finally have sample file with PGS: [https://www.w9y.de/sample-with-pgs.mkv](url). Exoplayer can play this file.

And while I was at it, I also used ffmpeg to convert the subtitles to DVBsub: [https://www.w9y.de/sample-with-dvbsub.mkv](url). The interesting bit about this one is that Exoplayer (demo app) can't play the file, but `ffplay` and `vlc` on my Laptop here can. I don't know anything about DVBsub, though...

icbaker commented:
> You can get it here, if you're interested to finally have sample file with PGS: https://www.w9y.de/sample-with-pgs.mkv. Exoplayer can play this file.

Amazing, thanks! I've added this to `MkvPlaybackTest` and sent it for review internally.

----

> Because we have to subtract the two bytes we just read ;-)

Oops, thanks for spotting and fixing that. I wasn't clear in my original comment but the test timed out both with and without my additional changes (and still times out with the latest fix too). It times out waiting for the player state to get to `IDLE`.

I tried playing the file in the demo app (thanks for the suggestion), it also doesn't ever stop - the UI says the file is `1193h02m47s` long. In VLC (on gLinux) I also see the same duration briefly, but playback does complete quickly (and so the UI gets hidden again).

`mediainfo` thinks it's specifically the VobSub track that is 1193h long [1].

And `mkvinfo` agrees:

```
|  + Simple
|   + Name: DURATION
|   + String: 1193:02:47.29500000
```

`mkvinfo -s` shows the single mega-long sample:

```
I frame, track 3, timestamp 00:00:00.000000000, duration 1193:02:47.295000000, size 10013, adler 0xa4986a81
```

Is it possible there was some time unit conversion problem when you were extracting and re-muxing?

----

[1]

```
$ mediainfo sample.mkv
General
Unique ID                                : 332219147464344373212352680433495465372 (0xF9EF14966859026D72420C95EBAA5D9C)
Complete name                            : sample.mkv
Format                                   : Matroska
Format version                           : Version 4
File size                                : 111 KiB
Duration                                 : 1193 h
Overall bit rate                         : 0 b/s
Frame rate                               : 612.245 FPS
Writing application                      : Lavf61.7.100
Writing library                          : Lavf61.7.100
ErrorDetectionType                       : Per level 1

Video
ID                                       : 1
Format                                   : AVC
Format/Info                              : Advanced Video Codec
Format profile                           : High@L5.2
Format settings                          : CABAC / 4 Ref Frames
Format settings, CABAC                   : Yes
Format settings, Reference frames        : 4 frames
Codec ID                                 : V_MPEG4/ISO/AVC
Duration                                 : 49 ms
Bit rate                                 : 14.3 Mb/s
Width                                    : 1 080 pixels
Height                                   : 720 pixels
Display aspect ratio                     : 3:2
Frame rate mode                          : Variable
Frame rate                               : 612.245 FPS
Color space                              : YUV
Chroma subsampling                       : 4:2:0
Bit depth                                : 8 bits
Scan type                                : Progressive
Bits/(Pixel*Frame)                       : 0.030
Stream size                              : 87.4 KiB (79%)
Writing library                          : x264 core 148 r2601 a0cd7d3
Encoding settings                        : cabac=1 / ref=3 / deblock=1:0:0 / analyse=0x3:0x113 / me=hex / subme=7 / psy=1 / psy_rd=1.00:0.00 / mixed_ref=1 / me_range=16 / chroma_me=1 / trellis=1 / 8x8dct=1 / cqm=0 / deadzone=21,11 / fast_pskip=1 / chroma_qp_offset=-2 / threads=12 / lookahead_threads=2 / sliced_threads=0 / nr=0 / decimate=1 / interlaced=0 / bluray_compat=0 / constrained_intra=0 / bframes=3 / b_pyramid=2 / b_adapt=1 / b_bias=0 / direct=1 / weightb=1 / open_gop=0 / weightp=2 / keyint=250 / keyint_min=25 / scenecut=40 / intra_refresh=0 / rc_lookahead=40 / rc=crf / mbtree=1 / crf=23.0 / qcomp=0.60 / qpmin=0 / qpmax=69 / qpstep=4 / ip_ratio=1.40 / aq=1:1.00
Default                                  : Yes
Forced                                   : No

Audio
ID                                       : 2
Format                                   : AC-3
Format/Info                              : Audio Coding 3
Commercial name                          : Dolby Digital
Codec ID                                 : A_AC3
Duration                                 : 1 s 9 ms
Bit rate mode                            : Constant
Bit rate                                 : 96.0 kb/s
Channel(s)                               : 1 channel
Channel layout                           : M
Sampling rate                            : 44.1 kHz
Frame rate                               : 28.711 FPS (1536 SPF)
Bit depth                                : 32 bits
Compression mode                         : Lossy
Stream size                              : 11.8 KiB (11%)
Service kind                             : Complete Main
Default                                  : Yes
Forced                                   : No
Dialog Normalization                     : -31 dB
dialnorm_Average                         : -31 dB
dialnorm_Minimum                         : -31 dB
dialnorm_Maximum                         : -31 dB

Text
ID                                       : 3
Format                                   : VobSub
Codec ID                                 : S_VOBSUB
Codec ID/Info                            : Picture based subtitle format used on DVDs
Duration                                 : 1193 h
Bit rate                                 : 162 kb/s
Frame rate                               : 0.000 (0/1000) FPS
Count of elements                        : 2
Stream size                              : 24.6 KiB (22%)
Writing library                          : Lavc61.19.100 dvdsub
Language                                 : English
Default                                  : Yes
Forced                                   : No
```

wischnow commented:
Oops. Indeed, ffmpeg messed up the track duration when converting PGS to Vobsub. I've just pushed the fixed version into the repository.

This also made me look at the DVBsub example again (see the URL above, the one Exoplayer couldn't play). That was messed up even more and the URL now points to the fixed version. Exoplayer still can't play it, though.

icbaker commented:
That's great, thanks for fixing the file - I've used this in the test and it now completes successfully.
